databuilder/codes.py

```diff
@@ -34,30 +39,84 @@ def _to_primitive_type(self):
 class BNFCode(BaseCode, system_id="bnf"):
     "Pseudo BNF"
 
+    regex = re.compile(
+        r"""
+        # Standard BNF code
+          # Chapter, Section, Paragraph, Sub-paragraph
+          [01][0-9]{6}
+          # Chemical
+          [0-9A-Z]{2}
+          # Product, strength-formulation, generic equivalent
+          ([A-Z][0-9A-Z]){3}
+        | # OR
+        # Appliances
+        2[0-3][0-9]{9}
+        """,
+        re.VERBOSE,
+    )
+
 
 class CTV3Code(BaseCode, system_id="ctv3"):
     "CTV3 (Read V3)"
 
-
-class DMDCode(BaseCode, system_id="dmd"):
-    "Dictionary of Medicines and Devices"
+    # Some of the CTV3 codes in OpenCodelists violate the below, either by having a
+    # leading dot or by starting with a tilde. However I have confirmed that, aside from
+    # a tiny handful of cases, these invalid codes are not used in the database.
+    regex = re.compile(
+        r"""
+        [0-9A-Za-z]{5}
+        | [0-9A-Za-z]{4}\.{1}
+        | [0-9A-Za-z]{3}\.{2}
+        | [0-9A-Za-z]{2}\.{3}
+        """,
+        re.VERBOSE,
+    )
 
 
 class ICD10Code(BaseCode, system_id="icd10"):
     "ICD-10"
 
+    regex = re.compile(r"[A-Z][0-9]{2,3}")
+
 
 class OPCS4Code(BaseCode, system_id="opcs4"):
     "OPCS-4"
 
+    # The documented structure requires three digits, and a dot between the 2nd and 3rd
+    # digit, but the codes we have in OpenCodelists omit the dot and sometimes have only
+    # two digits.
+    # https://en.wikipedia.org/wiki/OPCS-4#Code_structure
+    regex = re.compile(
+        r"""
+        # Uppercase letter excluding I
+        [ABCDEFGHJKLMNOPQRSTUVWXYZ]
+        [0-9]{2,3}
+        """,
+        re.VERBOSE,
+    )
+
 
 class Read2Code(BaseCode, system_id="readv2"):
     "Read V2"
 
+    # Syntactically equivalent to CTV3
+    regex = CTV3Code.regex
```

Comment:
Don't think this is right -- I'm pretty sure they can contain things like `XXXXX.00`.  But do we ever expose V2 codes anywhere?

Reply:
OK, I was just basing this off the Wikipedia page!
https://en.wikipedia.org/wiki/Read_code

But I think the only reason I added this as a type originally was that I just copied all the systems included in OpenCodelists. I don't think these are used anywhere in OpenSAFELY, so maybe the right thing to do is to remove the type?

Reply:
Let's remove the type completely!